cache/size_utils.go

```diff
@@ -0,0 +1,116 @@
+	// alive, we store pointers to all overflow buckets in hmap.overflow.
+	// Overflow is used only if key and value do not contain pointers.
+	// overflow[0] contains overflow buckets for hmap.buckets.
+	// overflow[1] contains overflow buckets for hmap.oldbuckets.
+	// The first indirection allows us to reduce static size of hmap.
+	// The second indirection allows to store a pointer to the slice in hiter.
+	overflow *[2]*[]*bmap
+}
+
+// A bucket for a Go map.
+type bmap struct {
+	// tophash generally contains the top byte of the hash value
+	// for each key in this bucket. If tophash[0] < minTopHash,
+	// tophash[0] is a bucket evacuation state instead.
+	tophash [bucketCnt]uint8
+	// Followed by bucketCnt keys and then bucketCnt values.
+	// NOTE: packing all the keys together and then all the values together makes the
+	// code a bit more complicated than alternating key/value/key/value/... but it allows
+	// us to eliminate padding which would be needed for, e.g., map[int64]int8.
+	// Followed by an overflow pointer.
+}
+*/
+
+const (
+	// MB is a short cut for 1024 * 1024.
+	MB = 1024 * 1024
+
+	// PtrSize is the number of bytes a pointer takes.
+	PtrSize = 4 << (^uintptr(0) >> 63) // stolen from runtime/internal/sys
+	// IntSize is the number of bytes an int or uint takes.
+	IntSize = 4 << (^uint(0) >> 63)
+
+	hmapStructSize = IntSize + // count int
+		1 + 1 + 2 + 4 + // flags, B, noverflow, hash0
+		PtrSize*3 + // buckets, oldbuckets, nevacuate
+		PtrSize + 2*PtrSize // overflow (estimate; not counting the slice)
+
+	bucketSizeWithoutIndirectPointerOverhead = 8 // tophash only
+)
+
+func mapKeyOrValueSizeWithIndirectPointerOverhead(rawSize int) int {
+	if rawSize > 128 {
+		// In Go maps, if key or value is larger than 128 bytes, a pointer type
+		// is used.
+		return rawSize + PtrSize
+	}
+	return rawSize
+}
+
+// StaticSizeOfMap provides a best-effort estimate of number of bytes that a
+// map takes in memory. It only includes statically sized content (i.e. struct,
+// array, int types, pointer address itself, slice/map's reference address
+// itself, etc.). If needed, dynamic sized stuff (slice/map content, pointer
+// content should be calculated separately by caller.
+func StaticSizeOfMap(
+	zeroValueKey, zeroValueValue interface{}, count int) (bytes int) {
+	return StaticSizeOfMapWithSize(int(reflect.TypeOf(zeroValueKey).Size()),
+		int(reflect.TypeOf(zeroValueValue).Size()), count)
+}
+
+// StaticSizeOfMapWithSize is a slightly more efficient version of
+// StaticSizeOfMap for when the caller knows the static size of key and value
+// without having to use `reflect`.
+func StaticSizeOfMapWithSize(
+	keyStaticSize, valueStaticSize int, count int) (bytes int) {
+	keySize := mapKeyOrValueSizeWithIndirectPointerOverhead(keyStaticSize)
+	valueSize := mapKeyOrValueSizeWithIndirectPointerOverhead(valueStaticSize)
+
+	B := math.Ceil(math.Log2(float64(count) / 6.5 /* load factor */))
```

Comment:
It's the reverse of how `map` allocates: https://github.com/keybase/kbfs/pull/1068/files#diff-ceba16b135943d7b0876b767cef0f52cR31